"Quality of life" improvements

goncalossilva opened the pull request:
This branch contains miscellaneous improvements, broken down into individual commits. These result from bugs / annoyances I've encountered while using the development version over the last few weeks.

Highlights:
* Improved inline preview, it does not need gnome-web-photo anymore and uses a WebView instead. Please verify that I've removed the dependency from flatpak correctly. I have never worked with flatpak before... and I just removed a file.
* Improved markup (eg. support for multi-line list items).
* Improved side-by-side preview when only one of the sides is scrollable.
* Improved stats counter.
* Moving from worker threads to worker processes, which improves general performance (particularly around UI) significantly. In particular, the experience when editing very large documents is much better. The GIL is a pain. :smile: 

I'd recommend reviewing commit by commit. I've taken care to explain each change(s) in each commit message, but I'm happy to discuss any of them. Thanks!

somas95 commented:
Hi, @goncalossilva! Sorry for my ausence. Nex week I finish my final exams, so I'll be able to review, merge and hopefully ship everything by then.

Switching to worker processes is a great idea IMHO. Another place where threadding/multiprocessing might improve things a lot is in the export dialog. Since we use pandoc for it I'd say it's an I/O bound module, so maybe threadding is better there, but I'm not sure

goncalossilva commented:
No problem at all @somas95! All of this can wait. Good luck for your finals!

wolfv commented:
all the best for the exams!

somas95 commented:
Ok, I'm starting to review the PR. I'll be commenting on the individual commits and commiting myself if I find issues
Again, thanks for your work, @goncalossilva! And sorry for the delay

goncalossilva commented:
Thanks for the feedback @somas95! Please keep it coming.

> As far as I know, the math one is wrong (I'm going to push a quick fix to it)

Sorry, I'm unsure how I missed this. 🤦‍♀

> Lists work way better than they used to, but the case
> * *something in italic*
> still fails

I didn't intend to add support for this particularly. In general, regexp is really not the way to parse markdown. It _mostly_ works, but it's a hack. We lose a lot of global context that markdown needs to be parsed correctly. This is why I was so excited about the AST, which unfortunately will have to wait.

But I can try. :smile: Will provide feedback.

somas95 commented:
>Sorry, I'm unsure how I missed this. :woman_facepalming:

It's really easy to not see this kind of things with REGEX. I'm building a set of unit tests to catch small errors (I found two that I didn't see back in the day)

It's not difficult, and I hope it'll help to depure the "hack" we have here. I know it could be improved (re.search only finds one match, but re.finditer makes a *lot* more difficult to catch some kind of errors since it yields an iterator, and well, iterators are kinda special when asserting things with them). Also I could have used a decorator, but I'm not sure all the tests will have the same structure (in fact I should check the "level" group in the header test). Anyway, those are unit tests, I don't want to over-engineer them

goncalossilva commented on uberwriter/markup_regex.py:
```diff
@@ -21,9 +21,9 @@
     r"^ {0,3}(?:> ?)+(?P<text>.+)", re.M)
 HEADER = re.compile(
-    r"^ {0,3}(?P<level>#{1,6}) (?P<text>[^\n]+)", re.M)
+    r"^ {0,3}(?P<level>#{1,6})(?P<text>[^\n]+)", re.M)
```
I might be missing something, but this change breaks headers. At least Commonmark and Pandoc's Markdown (the 2 variants I usually test) require a space between the `#` and the header text. The tests are incorrect as well as they lack this space.

goncalossilva commented on uberwriter/markup_regex.py:
```diff
@@ -21,9 +21,9 @@
+    r"^ {0,3}(?P<level>#{1,6})(?P<text>[^\n]+)", re.M)
 HEADER_UNDER = re.compile(
-    r"(?:^\n*|\n\n)(?P<text>[^\s].+)\n {0,3}[=\-]+(?:\s+?\n|$)")
+    r"(?:^\n*|\n\n)(?P<text>[^\s].+)\n[=\-]+(?: +?\n|$)")
```
The switch from `\s` to ` ` makes perfect sense, but the removal of the 3 optional spaces in the beginning is incorrect, at least in Commonmark and GitHub-Flavored Markdown which are arguably the most commonly used implementations. :blush:

There is a bug though: the 3 optional spaces are also allowed in the first line, not only the second one.

goncalossilva commented:
This approach is great! I get the search and finditer behave differently and that might not catch _some_ of the issues, but it's a really great start to have a test suite.

Having said this, I think we'll have to be very careful with what we add there. In particular, we have to chose one/two particular implementation to support properly in the inline formatting. We can't have it all. Just 2 simple examples:
* CommonMark, GitHub-Flavored Markdown and Pandoc's Markdown all require a space between the `#` and the header name. The other 2 don't.
* CommonMark and GitHub-Flavored Markdown allow up to 3 leading spaces on many block formats, such as quotes or multi-line headers. The other 3 don't.

There are _a lot_ of small inconsistencies between markdown implementations like these 2 examples. If we try to support them all, we'll do a sub-standard job overall. If we focus on 1/2, we might actually provide a solid experience. :smile: Of course, we should _try_ to support others as well, but one of them should be the priority for when a choice must be made like in the examples above.

Personally, I'd suggest CommonMark as the target. It's by far the strongest effort to standardize Markdown and provide cross-platform reference implementations. Could be the most widely in use as well due to this, I am not entirely sure.

somas95 commented:
The thing with TDD is that you have to be sure that the tests are correct, or the development will be wrong. I used windows tests as reference, assuming that they were correct (I remember that the "space between \# and the header text" grabbed my attention too)
I'll search for a detailed guide on markdown's syntax and be more careful with the commits to this

Also, if CommonMark is the most used flavour, not only we should use it as reference for the inline preview, but also default it for saving, exporting, etc

goncalossilva commented:
One of the great things about CommonMark is that they have a very comprehensive [spec](https://spec.commonmark.org/0.29/) and [online demo](https://spec.commonmark.org/0.29/). We could also explore the tests of any of the reference implementations.

goncalossilva commented:
The latest push contains some things that I had been already working before we had the discussion about the regexp. To prevent it from rotting, I've pushed it in.

All small changes. The largest would be improved undo/redo support for search and replace, which now counts as a single action. I hope this doesn't break anything as I was in the middle or revamping keyboard shortcuts for formatting, but I've stashed those changes for the time being (I've not had much time lately, so it'll take a while to be added) and briefly tested it. Should be fine.

As discussed before, I've also tweaked the regexp and tests (again, awesome idea to add them :raised_hands:) to prioritize CommonMark. It's just the last commit.

somas95 commented:
This looks great! Before merging I want to fix an issue, though: when opening a document it scrolls all the way down. Tomorrow I'll take a close look at it

goncalossilva commented:
Thanks for having a look @somas95! I have fixed the scrolling issue here: 43d442e8ef02f8895791dfea2d2fd79406151f49. Please have a look and let me know if you'd like to see any other change.

goncalossilva commented:
Have also rebased on the latest master and fixed the existing conflicts. Please check out cb3da0331eb4362dd50c4de1ec1edbce8dfc53d4. I picked `bin/uberwriter` over `uberwriter.in` but it's fine to do it the other way around if you prefer. :)

somas95 commented:
If you are ok with it, I'll revert back those two commits. Maybe we could keep the regexp expression for URIs...

goncalossilva commented:
Sure @somas95, as said before I am fine with either `uberwriter.in` or `bin/uberwriter`. My intent was to reduce duplication... We can absolutely keep `uberwriter.in` and remove `bin/uberwriter` instead!

> I thought in doing it this way, but this prevents the inline preview module from handling those links

Could you clarify which commit are you referring to?

goncalossilva commented:
Ah, I think I understand from the context. If it is what I think it is, it should be fixed in ec2f33e248059835e0065a8af12b6bffbcbee104. Instead of reverting to the previous workflow, I've added the `url` group to make the inline preview work. The reasoning is that these groups were added to provide semantic meaning and enable certain features seamlessly, like this one. Considering this, and that it applies to all other regexp, having both groups specified seems like a cleaner solution to me, but happy to discuss.

And `uberwriter.in` is back! :smile:

somas95 commented:
Ok, I tried to assign two groups to the same expression, without luck! Clever idea to encapsulate one into the other. I'm merging this!

goncalossilva commented:
Awesome! Thanks again for the thorough review!
